src/banking_stage.rs

```diff
@@ -10,75 +10,119 @@ use entry::Entry;
+        let tick_poh = poh.clone();
+        // Tick producer is a headless producer, so when it exits it should notify the banking stage.
+        // Since channel are not used to talk between these threads an AtomicBool is used as a
+        // signal.
+        let poh_exit = Arc::new(AtomicBool::new(false));
+        let banking_exit = poh_exit.clone();
+        // Single thread to generate entries from many banks.
+        // This thread talks to poh_service and broadcasts the entries once they have been recorded.
+        // Once an entry has been recorded, its last_id is registered with the bank.
+        let tick_producer = Builder::new()
+            .name("solana-banking-stage-tick_producer".to_string())
             .spawn(move || {
-                let poh_service = PohService::new(bank.last_id(), tick_duration.is_some());
-
-                let mut last_tick = Instant::now();
-
-                loop {
-                    let timeout =
-                        tick_duration.map(|duration| duration - (Instant::now() - last_tick));
-
-                    if let Err(e) = Self::process_packets(
-                        timeout,
-                        &bank,
-                        &poh_service,
-                        &verified_receiver,
-                        &entry_sender,
-                    ) {
-                        match e {
-                            Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
-                            Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
-                            Error::RecvError(_) => break,
-                            Error::SendError => break,
-                            _ => {
-                                error!("process_packets() {:?}", e);
-                                break;
-                            }
-                        }
-                    }
-                    if tick_duration.is_some() && last_tick.elapsed() > tick_duration.unwrap() {
-                        if let Err(e) = Self::tick(&poh_service, &bank, &entry_sender) {
-                            error!("tick() {:?}", e);
-                        }
-                        last_tick = Instant::now();
+                if let Err(e) = Self::tick_producer(tick_poh, config, &poh_exit) {
+                    match e {
+                        Error::SendError => (),
+                        _ => error!(
+                            "solana-banking-stage-tick_producer unexpected error {:?}",
+                            e
+                        ),
                     }
                 }
-                poh_service.join().unwrap();
+                debug!("tick producer exiting");
+                poh_exit.store(true, Ordering::Relaxed);
             }).unwrap();
-        (BankingStage { thread_hdl }, entry_receiver)
+
+        // Many banks that process transactions in parallel.
+        let mut thread_hdls: Vec<JoinHandle<()>> = (0..NUM_THREADS)
```

Comment:
How did any of this end up in this PR?

Reply:
@garious this is banking stage that can support an mt bank, which requires fixes in poh service/recorder.  as soon as bank is mt the banking stage can create more then one thread.  To me these are orthogonal, enabling the usage of mt bank and getting the rest of the system ready for it, vs actually making the bank mt